DcCore/DcCore/Helper/KeychainManager.swift

```diff
@@ -0,0 +1,83 @@
+import Foundation
+import Security
+
+public enum KeychainError: Error {
+    case noPassword
+    case unhandledError(message: String, status: OSStatus)
+}
+
+public class KeychainManager {
+    private typealias KcM = KeychainManager
+    private static let teamId = "8Y86453UA8"
+    private static let sharedKeychainGroup = "\(KcM.teamId).group.chat.delta.ios"
+
+    public static func getAccountSecret(accountID: Int) throws -> String {
+        do {
+            return try queryAccountSecret(id: accountID)
+        } catch KeychainError.noPassword {
+            return try addAccountSecret(id: accountID)
+        }
+    }
+
+    /**
+     * Deletes ALL secrets from keychain
+     * @return true if secrets have been deleted successfully or no secrets found
+     */
+    public static func deleteDBSecrets() -> Bool {
+        let query = [kSecClass as String: kSecClassGenericPassword,
+                     kSecAttrAccessGroup as String: KcM.sharedKeychainGroup as AnyObject
+                    ] as CFDictionary
+
+        let status = SecItemDelete(query)
+        return status == errSecSuccess || status == errSecItemNotFound
+    }
+
+    private static func createRandomPassword() -> String {
+        let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXZY1234567890"
+        return String((0..<36).map { _ in letters.randomElement()! })
+    }
+
+    private static func addAccountSecret(id: Int) throws -> String {
+        let keychainItemQuery = [
+          kSecValueData: createRandomPassword().data(using: .utf8)!,
+          kSecAttrAccount as String: "\(id)",
+          kSecClass: kSecClassGenericPassword,
+          kSecAttrAccessGroup as String: KcM.sharedKeychainGroup as AnyObject,
+        ] as CFDictionary
+
+        let status = SecItemAdd(keychainItemQuery, nil)
+        guard status == errSecSuccess else {
+            throw KeychainError.unhandledError(message: "Error adding secret for account \(id)",
+                                               status: status)
+        }
+        UserDefaults.shared?.set(true, forKey: UserDefaults.hasSavedKeyToKeychain)
+        return try queryAccountSecret(id: id)
+    }
+
+    private static func queryAccountSecret(id: Int) throws -> String {
+        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
+                                    kSecAttrAccount as String: "\(id)",
+                                    kSecMatchLimit as String: kSecMatchLimitOne,
+                                    kSecAttrAccessGroup as String: KcM.sharedKeychainGroup as AnyObject,
+                                    kSecReturnAttributes as String: true,
+                                    kSecReturnData as String: true]
+        var item: CFTypeRef?
+        let status = SecItemCopyMatching(query as CFDictionary, &item)
+        guard status != errSecItemNotFound else {
+            throw KeychainError.noPassword
```

Comment:
maybe we should return `null` if the account cannot be found instead of returning an error. that would make calling this function easier, see above.

Reply:
Generally, the idea was taken from: https://developer.apple.com/documentation/security/keychain_services/keychain_items/adding_a_password_to_the_keychain

We can replace the `.noPassword` error, but then we'll have to handle optionals everywhere we're calling this method instead, in order to distinct between the two general error cases (no password and every other possible error).

Reply:
Also here, thumbs up, if you still prefer returning `null`.

Reply:
i've overseen the explicit catch above, so, we can leave this as is, it is fine.